File: packages/domain/src/qr.ts
import type { Result } from './shared/result';
import { err, ok } from './shared/result';
import type { IsoDate } from './shared/time';
import { addSecondsIso, isAfter, secondsBetween } from './shared/time';

/**
 * DYNAMIC QR ≠ MEMBER ID: the QR encodes a short-lived, single-use access
 * credential, never a permanent identifier.
 */
export interface QrToken {
  token: string;
  memberId: string;
  issuedAt: IsoDate;
  expiresAt: IsoDate;
  consumedAt: IsoDate | null;
}

export function issueQrToken(args: {
  memberId: string;
  now: IsoDate;
  ttlSeconds: number;
  nonce: string;
}): QrToken {
  return {
    token: `qr_${args.nonce}`,
    memberId: args.memberId,
    issuedAt: args.now,
    expiresAt: addSecondsIso(args.now, args.ttlSeconds),
    consumedAt: null,
  };
}

export type QrTokenProblem = 'NOT_FOUND' | 'EXPIRED' | 'CONSUMED';

export function checkQrToken(
  token: QrToken | null | undefined,
  now: IsoDate,
): Result<QrToken, { reason: QrTokenProblem }> {
  if (!token) return err({ reason: 'NOT_FOUND' });
  if (token.consumedAt !== null) return err({ reason: 'CONSUMED' });
  if (isAfter(now, token.expiresAt)) return err({ reason: 'EXPIRED' });
  return ok(token);
}

export function qrSecondsRemaining(token: QrToken, now: IsoDate): number {
  return Math.max(0, Math.floor(secondsBetween(now, token.expiresAt)));
}

File: internal/domain/qr.go
package domain

import "time"

// QrToken is a short-lived, single-use access credential.
//
// It is deliberately NOT a member id: a screenshot of a member's code is
// useless within a minute, and each token opens the gate at most once.
type QrToken struct {
	Token      string     `json:"token"`
	MemberID   string     `json:"memberId"`
	IssuedAt   time.Time  `json:"issuedAt"`
	ExpiresAt  time.Time  `json:"expiresAt"`
	ConsumedAt *time.Time `json:"consumedAt"`
}

// IssueQrToken mints a credential valid for ttlSeconds. The nonce must be
// unguessable; the caller supplies it from a cryptographic source.
func IssueQrToken(memberID, nonce string, now time.Time, ttlSeconds int) QrToken {
	return QrToken{
		Token:     "qr_" + nonce,
		MemberID:  memberID,
		IssuedAt:  now,
		ExpiresAt: now.Add(time.Duration(ttlSeconds) * time.Second),
	}
}

// QrTokenProblem is why a token was rejected; empty means the token is good.
type QrTokenProblem string

const (
	QrOK       QrTokenProblem = ""
	QrNotFound QrTokenProblem = "NOT_FOUND"
	QrExpired  QrTokenProblem = "EXPIRED"
	QrConsumed QrTokenProblem = "CONSUMED"
)

// CheckQrToken validates a presented token. Consumption is checked before
// expiry so a replayed token always reads as CONSUMED, which is the more
// actionable signal at the door.
func CheckQrToken(token *QrToken, now time.Time) QrTokenProblem {
	switch {
	case token == nil:
		return QrNotFound
	case token.ConsumedAt != nil:
		return QrConsumed
	case now.After(token.ExpiresAt):
		return QrExpired
	default:
		return QrOK
	}
}

// QrSecondsRemaining is the countdown the member app renders around the code.
func QrSecondsRemaining(token QrToken, now time.Time) int {
	remaining := int(token.ExpiresAt.Sub(now).Seconds())
	if remaining < 0 {
		return 0
	}
	return remaining
}
